pygsti/algorithms/randomcircuit.py

```diff
@@ -2255,14 +2278,39 @@ def create_clifford_rb_circuit(pspec, clifford_compilations, length, qubit_label
 
     # Sample length+1 uniformly random Cliffords (we want a circuit of length+2 Cliffords, in total), compile
     # them, and append them to the current circuit.
-    for i in range(0, length + 1):
-
-        s, p = _symp.random_clifford(n, rand_state=rand_state)
-        circuit = _cmpl.compile_clifford(s, p, pspec,
-                                         clifford_compilations.get('absolute', None),
-                                         clifford_compilations.get('paulieq', None),
-                                         qubit_labels=qubit_labels, iterations=citerations, *compilerargs,
-                                         rand_state=rand_state)
+    num_native_gates = 0
+    for _ in range(0, length + 1):
+        if exact_compilation_key is not None:
+            # Deterministic compilation based on a provided clifford compilation
+            assert exact_compilation_key in clifford_compilations, \
+                    f"{exact_compilation_key} not provided in `clifford_compilations`"
+            
+            # Pick clifford
+            cidx = rand_state.randint(24**n)
+            lbl = _lbl.Label(f'C{cidx}', qubits)
+            
+            # Try to do deterministic compilation
+            try:
+                circuit = clifford_compilations[exact_compilation_key].retrieve_compilation_of(lbl)
+            except AssertionError:
+                raise ValueError(
+                    f"Failed to compile n-qubit Clifford 'C{cidx}'. Ensure this is provided in the " + \
+                    "compilation rules, or use a compilation algorithm to synthesize it by not " + \
+                    "specifying `exact_compilation_key`."
+                )
+        
+            # compute the symplectic rep of the chosen clifford
+            s, p = _symp.symplectic_rep_of_clifford_circuit(circuit, srep_cache)
```

Comment:
Note for if we ever want this to run quickly (we probably don't care) -- This is slightly absurdly way to do this, as given the index of the Clifford we should be able to easily compute s and p directly. However, there's only code right now to easily compute s (compute_symplectic_matrix(i, n)) and not p so there's no to-do I'm suggesting here.

Reply:
Agreed that this is not the best way to do this. What I actually wanted to do (and might still be useful in the future) was reuse the generate s/p code and then be able to look up the rule for compiling that Clifford from the CompilationRules. The CliffordCompRule object has some support for symplectic reps already, although not this reverse lookup (and I didn't want to restrict this to only taking CliffordCompRules).

So either way, I think we can revisit this when we update the symplectic tools to have either functionality (and I've left a TODO comment in the code noting that).